### PR TITLE
Mining QOL (Selective alloying,  TGUI'ed smelter)

### PR DESCRIPTION
**Pull request by MLGTASTICa**
## About The Pull Request
Lets the smelter have selective alloying , no more steel + plasteel from the same batch
the UI now uses TGUI
did some code QOL , separated MATERIALS from ORES

## Why It's Good For The Game
its better

## Testing
https://youtu.be/KOZDf3j7ovM

## Changelog
:cl:
refactor: Refactored the code of  the guild mining shuttle smelter and gave it a TGUI.
/:cl:

### CODE REVIEW

**Comment by DiaFRAME444** [comment]
Please let us just select what material we want to make instead of having to crossreference the wiki on what ore combination and refining technique makes what material.

**Comment by Mycah142** [comment]
> Please let us just select what material we want to make instead of having to crossreference the wiki on what ore combination and refining technique makes what material.

The video shows that's exactly what it does.

File: tgui/packages/tgui/interfaces/Processor.js [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  const {
+    materials_data = [],
+    alloy_data = [],
+    currently_alloying,
+    running,
+    sheet_rate,
+  } = data;
+  return data.machine ? (
+
+    <Window resizable>
```
Comment (LetterN) [inline tgui/packages/tgui/interfaces/Processor.js:17]: might want to set width and height here

File: tgui/packages/tgui/interfaces/Processor.js [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+      <Window.Content scrollable>
+        <Flex
+          frex-wrap="wrap"
+          >
+        <Flex.Item>
```
Comment (LetterN) [inline tgui/packages/tgui/interfaces/Processor.js:22]: indentation. eslint will complain, vscode should show the error too (its fixable)

File: tgui/packages/tgui/interfaces/Processor.js [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  ) : (<Window resizable>
+    No machine linked! There must be a material processor within 3 tiles for the wireless link to connect.
+    <Button
+      content="Attempt linking"
+      onClick={ () =>
+        act('machine_link')
+      }>
+      </Button>
+  </Window>
+    );
```
Comment (LetterN) [inline tgui/packages/tgui/interfaces/Processor.js:108]: do `if (!data.machine) return (this snippet)` so its more readable

**Comment by MLGTASTICa** [comment]
Since the head repository was deleted , maintainers need to remotely edit the PR after its been merged and replace the processor.js code with this , which is linter-compliat and styled properly
`import { useBackend } from '../backend';
import { Box, Button, Flex, Knob, LabeledList, Section } from '../components';
import { Window } from '../layouts';

export const Processor = (props, context) => {
  const { act, data } = useBackend(context);
  // Extract `health` and `color` variables from the `data` object.
  const {
    materials_data = [],
    alloy_data = [],
    currently_alloying,
    running,
    sheet_rate,
  } = data;
  if (data.machine) {
    return (
      <Window resizable>
        <Window.Content scrollable>
          <Flex frex-wrap="wrap">
            <Flex.Item>
              <Button
                content={running ? 'TURN OFF' : 'TURN ON'}
                onClick={() => act('set_running')}
              />
              <Box>
                <Knob
                  size={2}
                  minValue={5}
                  maxValue={30}
                  value={5}
                  unit="Sheets"
                  fillValue={sheet_rate}
                  content="Smelting rate"
                  step={1}
                  stepPixelSize={1}
                  onDrag={(e, value) =>
                    act('set_rate', {
                      sheets: value,
                    })
                  }
                />
                <br />
                <center>Melting Rate</center>
              </Box>
            </Flex.Item>
            <Flex.Item>
              <Section title="Loaded Materials">
                <LabeledList>
                  {materials_data.map((material) => (
                    <LabeledList.Item
                      key={material.name}
                      label={material.name}
                      buttons={
                        <Button
                          key={material.name}
                          content={material.current_action_string}
                          onClick={() =>
                            act('set_smelting', {
                              id: material.id,
                              action_type: material.current_action + 1,
                            })
                          }
                        />
                      }>
                      {material.amount}
                    </LabeledList.Item>
                  ))}
                </LabeledList>
              </Section>
              <Section title="Alloy Menu">
                <LabeledList>
                  {alloy_data.map((alloy) => (
                    <LabeledList.Item
                      key={alloy.name}
                      label={alloy.name}
                      buttons={
                        <Button
                          key={alloy.name}
                          content={alloy.name}
                          selected={alloy.name === currently_alloying}
                          onClick={() =>
                            act('set_alloying', {
                              id: alloy.name,
                            })
                          }
                        />
                      }
                    />
                  ))}
                </LabeledList>
              </Section>
            </Flex.Item>
          </Flex>
        </Window.Content>
      </Window>
      // Incase theres no machine
    );
  } else {
    <Window resizable>
      No machine linked! There must be a material processor within 3 tiles for
      the wireless link to connect.
      <Button content="Attempt linking" onClick={() => act('machine_link')} />
    </Window>;
  }
};
`

File: tgui/packages/tgui/interfaces/Processor.js [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+};
```
Comment (TheShown911) [inline tgui/packages/tgui/interfaces/Processor.js:109]:
```suggestion
import { useBackend } from '../backend';
import { Box, Button, Flex, Knob, LabeledList, Section } from '../components';
import { Window } from '../layouts';

export const Processor = (props, context) => {
  const { act, data } = useBackend(context);
  // Extract `health` and `color` variables from the `data` object.
  const {
    materials_data = [],
    alloy_data = [],
    currently_alloying,
    running,
    sheet_rate,
  } = data;
  if (data.machine) {
    return (
      <Window resizable>
        <Window.Content scrollable>
          <Flex frex-wrap="wrap">
            <Flex.Item>
              <Button
                content={running ? 'TURN OFF' : 'TURN ON'}
                onClick={() => act('set_running')}
              />
              <Box>
                <Knob
                  size={2}
                  minValue={5}
                  maxValue={30}
                  value={5}
                  unit="Sheets"
                  fillValue={sheet_rate}
                  content="Smelting rate"
                  step={1}
                  stepPixelSize={1}
                  onDrag={(e, value) =>
                    act('set_rate', {
                      sheets: value,
                    })
                  }
                />
                <br />
                <center>Melting Rate</center>
              </Box>
            </Flex.Item>
            <Flex.Item>
              <Section title="Loaded Materials">
                <LabeledList>
                  {materials_data.map((material) => (
                    <LabeledList.Item
                      key={material.name}
                      label={material.name}
                      buttons={
                        <Button
                          key={material.name}
                          content={material.current_action_string}
                          onClick={() =>
                            act('set_smelting', {
                              id: material.id,
                              action_type: material.current_action + 1,
                            })
                          }
                        />
                      }>
                      {material.amount}
                    </LabeledList.Item>
                  ))}
                </LabeledList>
              </Section>
              <Section title="Alloy Menu">
                <LabeledList>
                  {alloy_data.map((alloy) => (
                    <LabeledList.Item
                      key={alloy.name}
                      label={alloy.name}
                      buttons={
                        <Button
                          key={alloy.name}
                          content={alloy.name}
                          selected={alloy.name === currently_alloying}
                          onClick={() =>
                            act('set_alloying', {
                              id: alloy.name,
                            })
                          }
                        />
                      }
                    />
                  ))}
                </LabeledList>
              </Section>
            </Flex.Item>
          </Flex>
        </Window.Content>
      </Window>
      // Incase theres no machine
    );
  } else {
    <Window resizable>
      No machine linked! There must be a material processor within 3 tiles for
      the wireless link to connect.
      <Button content="Attempt linking" onClick={() => act('machine_link')} />
    </Window>;
  }
};
```

Comment (TheShown911) [reply tgui/packages/tgui/interfaces/Processor.js:109]: :mitosis:

File: .gitignore [hunk 5]
```diff
@@ -222,3 +222,4 @@ aux*.pdb
 
 # Screenshot tests
 /artifacts
+yarn.lock
```
Comment (TheShown911) [inline .gitignore:225]: Huh?

Comment (MLGTASTICa) [reply .gitignore:225]: update file from installing yarn , TGUI needs it.

File: tgui/packages/tgui/interfaces/Processor.js (hunk 4 above)
Comment (Pink-Chink) [inline tgui/packages/tgui/interfaces/Processor.js:109]:
```suggestion
import { useBackend } from '../backend';
import { Box, Button, Flex, Knob, LabeledList, Section } from '../components';
import { Window } from '../layouts';

export const Processor = (props, context) => {
  const { act, data } = useBackend(context);
  // Extract `health` and `color` variables from the `data` object.
  const {
    materials_data = [],
    alloy_data = [],
    currently_alloying,
    running,
    sheet_rate,
  } = data;
  if (data.machine) {
    return (
      <Window resizable>
        <Window.Content scrollable>
          <Flex frex-wrap="wrap">
            <Flex.Item>
              <Button
                content={running ? 'TURN OFF' : 'TURN ON'}
                onClick={() => act('set_running')}
              />
              <Box>
                <Knob
                  size={2}
                  minValue={5}
                  maxValue={30}
                  value={5}
                  unit="Sheets"
                  fillValue={sheet_rate}
                  content="Smelting rate"
                  step={1}
                  stepPixelSize={1}
                  onDrag={(e, value) =>
                    act('set_rate', {
                      sheets: value,
                    })
                  }
                />
                <br />
                <center>Melting Rate</center>
              </Box>
            </Flex.Item>
            <Flex.Item>
              <Section title="Loaded Materials">
                <LabeledList>
                  {materials_data.map((material) => (
                    <LabeledList.Item
                      key={material.name}
                      label={material.name}
                      buttons={
                        <Button
                          key={material.name}
                          content={material.current_action_string}
                          onClick={() =>
                            act('set_smelting', {
                              id: material.id,
                              action_type: material.current_action + 1,
                            })
                          }
                        />
                      }>
                      {material.amount}
                    </LabeledList.Item>
                  ))}
                </LabeledList>
              </Section>
              <Section title="Alloy Menu">
                <LabeledList>
                  {alloy_data.map((alloy) => (
                    <LabeledList.Item
                      key={alloy.name}
                      label={alloy.name}
                      buttons={
                        <Button
                          key={alloy.name}
                          content={alloy.name}
                          selected={alloy.name === currently_alloying}
                          onClick={() =>
                            act('set_alloying', {
                              id: alloy.name,
                            })
                          }
                        />
                      }
                    />
                  ))}
                </LabeledList>
              </Section>
            </Flex.Item>
          </Flex>
        </Window.Content>
      </Window>
      // Incase theres no machine
    );
  } else {
    <Window resizable>
      No machine linked! There must be a material processor within 3 tiles for
      the wireless link to connect.
      <Button content="Attempt linking" onClick={() => act('machine_link')} />
    </Window>;
  }
};
```

File: code/__DEFINES/materials.dm [hunk 6]
```diff
@@ -64,3 +64,26 @@
 #define PLASTEEL_COLOUR "#8e9799"
 //Color to exactly match concept art #b0bcbe
 //I made it slightly darker than that, because the original color was too bright and causes eyestrain imo
+
+
+#define ORE_CARBON "o_carbon"
```
Comment (ValoTheValo) [inline code/__DEFINES/materials.dm:69]: this broke rock ore overlays

**Pull request merged**